### PR TITLE
M3-5190: React Query for Object Storage

### PR DESCRIPTION
**Pull request by bnussman**
## Description

React Query for Object Storage 

- [x] Clusters
- [x] Buckets (may need to rework to use old per-cluster logic)
- [x] Bucket - Create
- [x] Bucket - Delete
- [x] Bucket Contents (will need to refactor to be a functional component)
- [x] Search Buckets

## How to test

- Test **all** Object Storage Functionality
- Use the MSW to see how `/object-storage/buckets` handles a cluster being offline

### CODE REVIEW

File: packages/manager/src/features/ObjectStorage/BucketLanding/BucketLanding.tsx [hunk 1]
```diff
@@ -186,52 +199,64 @@ export const BucketLanding: React.FC<CombinedProps> = (props) => {
...
     return <RenderLoading />;
   }
 
-  const allBucketRequestsFailed =
-    bucketErrors?.length === objectStorageClusters.entities.length;
-
-  // Show a general error state if all the bucket requests failed.
-  if (allBucketRequestsFailed || objectStorageClusters.error) {
-    return <RenderError data-qa-error-state />;
-  }
```
Comment (bnussman) [inline packages/manager/src/features/ObjectStorage/BucketLanding/BucketLanding.tsx:216]: Re-add this type of error state

**Review by codyfinn — CHANGES_REQUESTED** [review]
Looks good overall. The way we are searching through the paginated data seems pretty inefficient. It would be nice to do so in a more performant way.

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx [hunk 2]
```diff
@@ -279,53 +156,64 @@ export class BucketDetail extends React.Component<CombinedProps, State> {
...
+
+        return {
+          is_truncated: page.is_truncated,
+          next_marker: page.next_marker,
+          data: updatedData,
+        } as ObjectStorageObjectListResponse;
+      }) || [];
+
+    updateStore(newPagesArray);
+  };
```
[the remark this answers is not in the document]
Comment (bnussman) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:199]: Yes, but it's my understanding that two objects can't have the same name. Is my understanding correct?

[the remark this answers is not in the document]
Comment (codyfinn) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:199]: Yeah looking closer at how the API handles it that seems to be the case. I still think finding a way to not continue searching after you have found the entry to delete would be worth it though.

**Review by HanaXu — COMMENTED** [review]
Can we merge develop in and resolve conflicts?

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (codyfinn) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:199]:
```suggestion
  const removeOne = (objectName: string) => {
    const newPagesArray = produce(data?.pages, (draft) => {
      let objectIndex = -1;
      const pageIndex = draft?.findIndex((thisPage) => {
        const _objectIndex = thisPage.data.findIndex(
          (object) => object.name === objectName
        );
        objectIndex = _objectIndex !== -1 ? _objectIndex : -1;
        return _objectIndex !== -1;
      });
      if (draft && pageIndex !== undefined && pageIndex !== -1) {
        draft[pageIndex].data.splice(objectIndex);
      }
    });

    updateStore(newPagesArray ?? []);
  };
```
Just a rough draft of what that could look like

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx [hunk 3]
```diff
@@ -366,185 +256,156 @@ export class BucketDetail extends React.Component<CombinedProps, State> {
...
+              <Typography variant="subtitle2" className={classes.footer}>
+                Showing all {numOfDisplayedObjects} items
+              </Typography>
+            )}
+            <ConfirmationDialog
+              open={deleteObjectDialogOpen}
+              onClose={closeDeleteObjectDialog}
+              title={
+                objectToDelete
+                  ? `Delete  ${truncateMiddle(displayName(objectToDelete))}`
```
Comment (HanaXu) [inline packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:362]: ``Delete ${truncateMiddle(displayName(objectToDelete))}``

File: packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx [hunk 4]
```diff
@@ -97,39 +100,6 @@ export const ObjectStorageLanding: React.FC<CombinedProps> = (props) => {
     createOrEditDrawer.open();
   };
 
-  const { _isRestrictedUser, accountSettings } = useAccountManagement();
```
Comment (HanaXu) [inline packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx:100]: There seems to be some kind of regression for the restricted user `Create Access Key` drawer

This branch:
image

Prod:
image

**Review by HanaXu — CHANGES_REQUESTED** [review]
The error loading notice is a bit big, can we reduce the margin and maybe remove the last sentence as it's a bit redundant?

image

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx (hunk 3 above)
Comment (bnussman) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:362]: What do you mean by this? @HanaXu

File: packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx (hunk 4 above)
Comment (bnussman) [reply packages/manager/src/features/ObjectStorage/ObjectStorageLanding.tsx:100]: Resolved by a59f125

File: packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx (hunk 3 above)
Comment (bnussman) [reply packages/manager/src/features/ObjectStorage/BucketDetail/BucketDetail.tsx:362]: Didn't even notice, thank you!

**Comment by bnussman** [comment]
@HanaXu Would you be ok with holding off on those UI changes to keep the scope of this PR focused?

File: packages/manager/src/mocks/serverHandlers.ts [hunk 5]
```diff
@@ -490,7 +490,41 @@ export const handlers = [
...
+          (page - 1) * pageSize + pageSize
+        ),
+        page,
+        pages: Math.ceil(buckets.length / pageSize),
+        results: buckets.length,
+      })
+    );
+  }),
+  rest.get('*/object-storage/buckets', (req, res, ctx) => {
+    const buckets = objectStorageBucketFactory.buildList(10);
```
Comment (DevDW) [inline packages/manager/src/mocks/serverHandlers.ts:527]: Why are we mocking the GET request 4 times here?

**Review by DevDW — COMMENTED** [review]
Nice work! The functionality and Redux cleanup are looking good overall, left a few comments

File: packages/manager/src/mocks/serverHandlers.ts (hunk 5 above)
Comment (bnussman) [reply packages/manager/src/mocks/serverHandlers.ts:527]: When we fetch Object Storage Buckets, we make an API request for each cluster. We do this so we know if any of the clusters are offline. I mocked 4 times so I could simulate some failures to test clusters that error out. I can remove the erroring mocks prior to merge.

File: packages/manager/src/components/PrimaryNav/PrimaryNav.tsx [hunk 6]
```diff
@@ -199,10 +169,6 @@ export const PrimaryNav: React.FC<Props> = (props) => {
             '/object-storage/access-keys',
           ],
           icon: <Storage />,
-          prefetchRequestFn: () => requestObjectStorageBuckets(clusterIds),
-          prefetchRequestCondition:
-            !objectStorageBuckets.loading &&
-            objectStorageBuckets.lastUpdated === 0,
```
[the remark this answers is not in the document]
Comment (bnussman) [reply packages/manager/src/components/PrimaryNav/PrimaryNav.tsx:205]: Sounds good! Will re-add prefetching

**Review by HanaXu — APPROVED** [review]
Object Storage LGTM, great job! 🚀

**Review by codyfinn — APPROVED** [review]

**Pull request merged**